hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java

```diff
@@ -0,0 +1,167 @@
+        props.getInteger(Config.NUM_INSTANTS_PER_FETCH, Config.DEFAULT_NUM_INSTANTS_PER_FETCH);
+    boolean readLatestOnMissingCkpt =
+        props.getBoolean(
+            Config.READ_LATEST_INSTANT_ON_MISSING_CKPT,
+            Config.DEFAULT_READ_LATEST_INSTANT_ON_MISSING_CKPT);
+
+    // Use begin Instant if set and non-empty
+    Option<String> beginInstant =
+        lastCkptStr.isPresent()
+            ? lastCkptStr.get().isEmpty() ? Option.empty() : lastCkptStr
+            : Option.empty();
+
+    Pair<String, String> instantEndpts =
+        IncrSourceHelper.calculateBeginAndEndInstants(
+            sparkContext, srcPath, numInstantsPerFetch, beginInstant, readLatestOnMissingCkpt);
+
+    if (instantEndpts.getKey().equals(instantEndpts.getValue())) {
+      LOG.warn("Already caught up. Begin Checkpoint was :" + instantEndpts.getKey());
+      return Pair.of(Option.empty(), instantEndpts.getKey());
+    }
+
+    // Do Incr pull. Set end instant if available
+    DataFrameReader reader =
+        sparkSession
+            .read()
+            .format("org.apache.hudi")
+            .option(
+                DataSourceReadOptions.QUERY_TYPE_OPT_KEY().key(),
+                DataSourceReadOptions.QUERY_TYPE_INCREMENTAL_OPT_VAL())
+            .option(
+                DataSourceReadOptions.BEGIN_INSTANTTIME_OPT_KEY().key(), instantEndpts.getLeft())
+            .option(
+                DataSourceReadOptions.END_INSTANTTIME_OPT_KEY().key(), instantEndpts.getRight());
+
+    Dataset<Row> source = reader.load(srcPath);
+
+    // Extract distinct file keys from cloud meta hoodie table
+    final List<Row> cloudMetaDf =
+        source.filter("s3.object.size > 0").select("s3.bucket.name", "s3.object.key").distinct().collectAsList();
+
+    // Create S3 paths
+    List<String> cloudFiles = new ArrayList<>();
+    for (Row row : cloudMetaDf) {
+      String bucket = row.getString(0);
+      String key = row.getString(1);
+      String filePath = "s3://" + bucket + "/" + key;
+      cloudFiles.add(filePath);
+    }
+    String pathStr = String.join(",", cloudFiles);
+
+    // log.info("Final Schema from Source is :" + src.schema());
+    return Pair.of(Option.of(fromFiles(pathStr)), instantEndpts.getRight());
+  }
+
+  /** Function to create Dataset from files. */
+  private Dataset<Row> fromFiles(String pathStr) {
+    return sparkSession.read().parquet(pathStr.split(","));
+  }
+
+  /** Config Class. {@link Config} */
+  protected static class Config {
+
+    /** {@value #HOODIE_SRC_BASE_PATH} is the base-path for the source Hoodie table. */
+    private static final String HOODIE_SRC_BASE_PATH =
+        "hoodie.deltastreamer.source.hoodieincr.path";
+
+    /**
+     * {@value #NUM_INSTANTS_PER_FETCH} allows the max number of instants whose changes can be
+     * incrementally fetched.
+     */
+    private static final String NUM_INSTANTS_PER_FETCH =
+        "hoodie.deltastreamer.source.hoodieincr.num_instants";
```

Comment:
Let's rename this to `hoodie.deltastreamer.source.hoodieincr.num.instants` or `hoodie.deltastreamer.source.hoodieincr.instants`. Idea is to not have underscore, if we need underscore we can replace it by `.` as well.